clean up gossip key handling

tbg opened the pull request:
keys, patterns and their creating methods are now located in a central
place.
nodeID and storeID are converted to decimal strings.

I've also added a commit upping the test timeout to 15s. Unfortunately the `kv` test is getting rather slow and should be refactored.

tbg committed:
increase default test timeout to 15s

this is mostly due to the ./kv test clocking in at around
10s most of the time (failing perhaps 65% of the time on
my relatively fast (or so I thought) machine).

We should factor out the txn_correctness_test. I had tried
that, but could not complete it due to import issues.
The correctness test alone takes around 5s.

petermattis commented on storage/store.go:
```diff
@@ -472,7 +472,7 @@ func (s *Store) Start() error {
 		s.gossip.RegisterCallback(gossip.KeyConfigZone, s.configGossipUpdate)
 		// Callback triggers on capacity gossip from all stores.
-		capacityRegex := fmt.Sprintf("%s.*", gossip.KeyMaxAvailCapacityPrefix)
+		capacityRegex := gossip.MakePrefixPattern(gossip.KeyMaxAvailCapacityPrefix)
```
Not sure if this is any clearer. I didn't mind the `fmt.Sprintf` or you could just concatenate the strings:

```
  capacityRegex := gossip.KeyMaxAvailCapacityPrefix + ".*"
```

tschottdorf replied:
I agree that it doesn't look much clearer now, but I think it beats manually building regexes in a number of places that can only go up. I think without this we might just see weird bugs related to the fact that gossip keys can be prefixes of each other and then it matters that your regexes are solid (think node-1 vs node-count).

petermattis replied:
Ok, though `MakePrefixPattern` is only used in one location right now. I don't feel too strongly about this.

petermattis commented on proto/config.go:
```diff
@@ -31,6 +32,12 @@ import (
 
+// String implements the fmt.Stringer interface.
+// It is used to format the ID for use in Gossip keys.
+func (n NodeID) String() string {
```
This might have unintended interactions with `fmt` (e.g. if "%08d" is ever used to format a `NodeID`). Probably only affects log messages, but just wanted to point it out.

tschottdorf replied:
I'm confused - do you expect problems with something like this? http://play.golang.org/p/4cj8PIaNC7

petermattis replied:
Yeah, I would have guessed that didn't work properly. Perhaps it does.

petermattis commented:
LGTM

spencerkimball commented on run/local-cluster.sh:
```diff
@@ -165,7 +165,7 @@ for ATTEMPT in $(seq 1 $MAX_WAIT); do
     GOSSIP=$(curl --noproxy '*' -s $GOSSIP_URL)
     for j in $(seq 1 $((2*NODES))); do
-      if [[ ! -z $(echo $GOSSIP | grep "node-$j") ]]; then
+      if [[ ! -z $(echo $GOSSIP | grep -E "node.$j") ]]; then
```
I thought it was a colon separator. With grep -E, you're allowing any character between "node" and the node ID. Why not just grep "node:$j"?

tschottdorf replied:
Seemed reasonably specific. I can hard-code the ':' though.

spencerkimball commented:
LGTM

tbg committed:
clean up gossip key handling

keys, patterns and their creating methods are now located in a central
place.
nodeID and storeID are converted to decimal strings.
